scripts/end-to-end/extract-testcases.py

```diff
@@ -0,0 +1,284 @@
+            "asString(", "").replace("fromHex(", "0x").replace("\"true\"", "true").replace("\"false\"", "false").strip()
+        result = ""
+        if string.startswith(","):
+            string = string[1:].strip()
+        if string.startswith("0x") and string.endswith("\""):
+            string = string.replace("\"", "")
+        if string:
+            r = ""
+            for arg in string.split(","):
+                arg = arg.strip()
+                if (arg == "true"):
+                    result += "true, "
+                elif (arg == "false"):
+                    result += "false, "
+                else:
+                    try:
+                        evaluated = eval(arg)
+                        if isinstance(evaluated, int):
+                            if string.find("0x") >= 0:
+                                r = hex(evaluated)
+                            else:
+                                r = int(evaluated)
+                        elif isinstance(evaluated, str):
+                            r = arg
+                        result += str(r) + ", "
+                    except (SyntaxError, NameError, TypeError, ZeroDivisionError):
+                        self.extractable = False
+                        result += "?, "
+            return result[:-2]
+        return ""
+
+    def create_call(self, signature, parameters, expectations, comment, withValue=False):
+        result = signature
+        value = ""
+        if withValue:
+            if parameters.find(",") == -1:
+                value = parameters
+                parameters = ''
+        if parameters:
+            result += ": "
+            result += parameters
+        if value:
+            result += ", " + str(value) + " wei"
+        result += " -> "
+        if expectations:
+            result += expectations
+        if comment:
+            result += " # " + comment + " #"
+        return result
+
+    def create_isoltest_call(self, line, test_comment):
+        search = re.search(r'ABI_CHECK\(callContractFunction\(\"(.*)\",((.|\s)*)\), encodeArgs\(((.|\s)*)\)\);', line,
+                           re.M | re.I)
+        if search:
+            result = self.create_call(search.group(1).strip(), self.eval_and_correct(search.group(2)),
+                                      self.eval_and_correct(search.group(4)), test_comment)
+            return result
+
+        search = re.search(
+            r'ABI_CHECK\(callContractFunctionWithValue\(\"(.*)\",((.|\s)*)\), encodeArgs\(((.|\s)*)\)\);', line,
+            re.M | re.I)
+        if search:
+            result = self.create_call(search.group(1).strip(), self.eval_and_correct(search.group(2)),
+                                      self.eval_and_correct(search.group(4)), test_comment, True)
+            return result
+
+        search = re.search(r'ABI_CHECK\(callContractFunction\(\"(.*)\",((.|\s)*)\), fromHex\("((.|\s)*)"\)\);', line,
```

Comment:
Wouldn't it be more readable to first extract the `callContractFunction` part and then the second argument to `ABI_CHECK`? You could even use `r'ABI_CHECK\(callContractFunction(WithValue)?\(...`

And also please use non-greedy asterisks (`*?`) everywhere.

Reply:
yes.. the main thing was that I didn't wanted to change the initial extractor script too much, because we will need to regenerate everything again.. I totally agree that the extractor is super hacky..